Cortex-R mpu fix

bbolen opened the pull request:
@ibirnbaum This adds a patch for renaming the cortex-r kconfig to include "aarch32" that stephanosio suggested in your cortex-a pull request.

Commit:
arch: arm: core: aarch32: Fix the syscall design for Cortex-R

When calling a syscall, the SVC routine will now elevate the thread to
privileged mode and exit the SVC setting the return address to the
syscall handler.  When the thread is swapped back in, it will be running
z_do_arm_syscall in system mode.  That function will run the syscall
then automatically return the thread to usr mode.

This allows running the syscall in sys mode on a thread so that we can
use syscalls that sleep without doing unnatural things.  The previous
implementation would enable interrupts while still in the SVC call and
do weird things with the nesting count.  An interrupt could happen
during this time when the syscall was still in the exception state, but
the nested count had been decremented too soon.  Correctness of the
nested count is important for future floating point unit work.

The Cortex-R behavior now matches that of Cortex-M.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
arch: arm: core: aarch32: Change Cortex-R config check

Replace CONFIG_CPU_CORTEX_R with CONFIG_ARMV7_R since it is clearer with
respect to the difference between v7 and v8 Cortex-R.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

tejlmand commented:
> @ibirnbaum This adds a patch for renaming the cortex-r kconfig to include "aarch32" that stephanosio suggested in your cortex-a pull request.

Could you please link the refered PR ?
Also, which commits of this PR should be reviewed ?
If this PR is not intended for review, please put it in draft and mention so in the PR message.

bbolen commented:
> > @ibirnbaum This adds a patch for renaming the cortex-r kconfig to include "aarch32" that stephanosio suggested in your cortex-a pull request.
> 
> Could you please link the refered PR ? Also, which commits of this PR should be reviewed ? If this PR is not intended for review, please put it in draft and mention so in the PR message.

https://github.com/zephyrproject-rtos/zephyr/pull/36983#discussion_r738358044

I put in the DNM b/c I was gone for 2 weeks and wouldn't be able to respond to anything.  I'll remove it.  All 3 commits should be reviewed.

Commit:
arch: arm: core: aarch32: Use cmsis functions

These functions help the code to be more self-documenting.  Use them to
make the code's intent clearer.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

bbolen commented:
@ibirnbaum Any chance you could take a look at this?

ibirnbaum commented on arch/arm/core/aarch32/userspace.S:
```diff
@@ -574,18 +553,21 @@ dispatch_syscall:
     pop {r2, r3}
-#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE) \
+    || defined(CONFIG_CPU_CORTEX_R)
```
Could this instance of CONFIG_CPU_CORTEX_R be replaced by CONFIG_ARMV7_R as well?

ibirnbaum reviewed (approved):
Looks good to me.

bbolen commented:
@stephanosio Can you take a look at this?

stephanosio commented on include/arch/arm/aarch32/cortex_a_r/cpu.h:
```diff
@@ -14,7 +14,7 @@
 /*
  * SCTRL register bit assignments
  */
-#define SCTRL_MPU_ENABLE	(1 << 0)
+#define SCTLR_MPU_ENABLE	(1 << 0)
```
would be nice to update the comment as well.

stephanosio reviewed (approved):
I have not done a comprehensive analysis on the Cortex-R syscall architecture change; but, it looks reasonable at first glance.
